src/views/common/SortableTableWidget/SortableTableWidgetHeader.tsx

```diff
@@ -0,0 +1,145 @@
+        // Skipping undefined columns prevents a crash, but sorting results might not be completely honored if we're sorting on a column that
+        // has ties, and the appropriate tie-breaking field doesn't exist on the table instance actively doing the sorting.
+        //    Practically speaking, this can't happen at present, because it can only affect the SelectUnitsWidget, which only has a (unique)
+        // _unitId column, so there can never be any ties to break.
+        if (column) {
+            _draft.sort((a, b) => {
+                const aValue = (a.data[columnName] || {}).sortValue
+                const bValue = (b.data[columnName] || {}).sortValue
+                return rule.sortAscending ? column.sort(aValue, bValue) : column.sort(bValue, aValue)
+            })
+        }
+    })
+    return _draft.map(row => row.rowIdNumeric)
+}
+
+const getColumnNameDict = (columns: SortableTableWidgetColumn[]): Map<string, SortableTableWidgetColumn> => {
+    const map = new Map<string, SortableTableWidgetColumn>()
+    columns.forEach(c => map.set(c.columnName, c))
+    return map
+}
+
+const mapColumnsToHeaderInfo = (columns: SortableTableWidgetColumn[], primarySortRule?: SortingRule): ColumnHeaderInfo[] => {
+    return columns.map((c) => ({
+        name: c.columnName,
+        tooltip: c.tooltip,
+        label: c.label,
+        isCalculating: c.calculating || false,
+        isPrimarySort: c.columnName === primarySortRule?.columnName,
+        isAscendingSort: primarySortRule?.sortAscending || false
+    }))
+}
+
+const SortableTableWidgetHeaderRow: FunctionComponent<HeaderRowProps> = (props) => {
+    const { columns, selectionDispatch, primarySortRule, allRowSelectionStatus, rowSorterCallback, selectionDisabled } = props
+    const columnsMap = useMemo(() => getColumnNameDict(columns), [columns])
+    const columnHeaders = useMemo(() => mapColumnsToHeaderInfo(columns, primarySortRule), [columns, primarySortRule])
+
+    const toggleSelectAllCallback = useCallback(() => selectionDispatch({type: TOGGLE_SELECT_ALL}), [selectionDispatch])
+    // We don't want the headers to have to worry about the row contents, but it makes sense for sorting logic to live with
+    // the columns. So we have the SortableTableWidget parent give us a memoized function that has the rows in it already.
+    // Then we use our own callback that changes only when that, or the columns map, changes; this one takes a set of rules
+    // and uses the wrapped function (with the rows) and the columns (which we know here) to call the sortRows function
+    // defined above. The result is a function mapping the set of rules to the ID ordering. That function is suitable for
+    // passing to the reducer.
+    const sortCallback = useCallback((rules: SortingRule[]) => rowSorterCallback(columnsMap)(rules), [rowSorterCallback, columnsMap])
```

Comment:
So above we defined a (static, not part of this component & thus not subject to re-rendering) `sorterCallbackWrapper` function that takes a set of rows and columns and returns a function mapping rules onto sorted numbers.

We also have a prop, `rowSorterCallback`, that maps columns to a function taking rules and yielding numbers. This prop is satisfied by our parent component, `SortableTableWidget`, which has a `useCallback` hook (updating only on row set changes) to wrap the table rows (with all their data fields) into the `sorterCallbackWrapper` call. That provides a function which takes the set of columns only, and yields a sorting callback (function from rules to sorted ids).

Basically, [SortableTableWidget](https://github.com/magland/spikesortingview/pull/41/files#diff-4f473aee87573e6a7158b1bedbc9d3450045e3811e5823156db47e036b896186R13) memoized the actual rowset and gave us a curried function of columns to (rules to results). Then this line memoizes & curries the column set, yielding a not-constantly-replaced function from rules to results, which we give as a callback function to the reducer in line 96.